File: ios/lib/features/auth/data/services/auth_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/auth_models.dart';

class AuthService {
  static const String _baseUrl = 'http://localhost:8080/api/v1';
  static const String _tokenKey = 'auth_token';
  static const String _userKey = 'user_profile';
  
  // Send magic link to email
  Future<AuthResponse> sendMagicLink(String email) async {
    final response = await http.post(
      Uri.parse('$_baseUrl/auth/magic-link'),
      headers: {
        'Content-Type': 'application/json',
      },
      body: json.encode({'email': email}),
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return AuthResponse(
        message: data['message'] ?? '매직링크가 전송되었습니다.',
        success: data['success'] ?? true,
      );
    } else {
      final error = json.decode(response.body);
      throw Exception(error['message'] ?? '매직링크 전송에 실패했습니다.');
    }
  }
  
  // Verify magic link token
  Future<TokenVerifyResponse> verifyMagicLink(String token) async {
    final response = await http.get(
      Uri.parse('$_baseUrl/auth/verify?token=$token'),
      headers: {
        'Content-Type': 'application/json',
      },
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      
      // Save token and user data
      await _saveAuthData(
        data['token'],
        data['user'],
      );
      
      return TokenVerifyResponse(
        token: data['token'],
        isNewUser: data['is_new_user'] ?? false,
        user: UserProfile.fromJson(data['user']),
      );
    } else {
      final error = json.decode(response.body);
      throw Exception(error['message'] ?? '인증에 실패했습니다.');
    }
  }
  
  // Get current user profile
  Future<UserProfile?> getCurrentUser() async {
    try {
      final token = await getToken();
      if (token == null) return null;
      
      final response = await http.get(
        Uri.parse('$_baseUrl/auth/profile'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );
      
      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return UserProfile.fromJson(data['user']);
      }
      return null;
    } catch (e) {
      print('Get current user error: $e');
      return null;
    }
  }
  
  // Logout user
  Future<void> logout() async {
    try {
      final token = await getToken();
      if (token != null) {
        await http.post(
          Uri.parse('$_baseUrl/auth/logout'),
          headers: {
            'Content-Type': 'application/json',
            'Authorization': 'Bearer $token',
          },
        );
      }
    } catch (e) {
      print('Logout error: $e');
    } finally {
      await _clearAuthData();
    }
  }
  
  // Check if user is authenticated
  Future<bool> isAuthenticated() async {
    final token = await getToken();
    if (token == null) return false;
    
    try {
      final user = await getCurrentUser();
      return user != null;
    } catch (e) {
      return false;
    }
  }
  
  // Get stored auth token
  Future<String?> getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_tokenKey);
  }
  
  // Get stored user profile
  Future<UserProfile?> getStoredUser() async {
    final prefs = await SharedPreferences.getInstance();
    final userJson = prefs.getString(_userKey);
    if (userJson != null) {
      return UserProfile.fromJson(json.decode(userJson));
    }
    return null;
  }
  
  // Save authentication data
  Future<void> _saveAuthData(String token, Map<String, dynamic> userData) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_tokenKey, token);
    await prefs.setString(_userKey, json.encode(userData));
  }
  
  // Clear authentication data
  Future<void> _clearAuthData() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_tokenKey);
    await prefs.remove(_userKey);
  }
}

File: android/app/src/main/java/com/signal/app/features/auth/data/services/AuthService.kt
package com.signal.app.features.auth.data.services

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.google.gson.Gson
import com.signal.app.features.auth.data.models.AuthRequest
import com.signal.app.features.auth.data.models.AuthResponse
import com.signal.app.features.auth.data.models.TokenVerifyResponse
import com.signal.app.features.auth.data.models.UserProfile
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "auth_preferences")

@Singleton
class AuthService @Inject constructor(
    private val authApi: AuthApi,
    @ApplicationContext private val context: Context,
    private val gson: Gson
) {
    companion object {
        private val TOKEN_KEY = stringPreferencesKey("auth_token")
        private val USER_KEY = stringPreferencesKey("user_profile")
    }

    // Send magic link to email
    suspend fun sendMagicLink(email: String): AuthResponse {
        val response = authApi.sendMagicLink(AuthRequest(email))
        if (response.isSuccessful) {
            return response.body() ?: throw Exception("매직링크 전송에 실패했습니다.")
        } else {
            throw Exception(response.errorBody()?.string() ?: "매직링크 전송에 실패했습니다.")
        }
    }

    // Verify magic link token
    suspend fun verifyMagicLink(token: String): TokenVerifyResponse {
        val response = authApi.verifyMagicLink(token)
        if (response.isSuccessful) {
            val result = response.body() ?: throw Exception("인증에 실패했습니다.")
            
            // Save token and user data
            saveAuthData(result.token, result.user)
            
            return result
        } else {
            throw Exception(response.errorBody()?.string() ?: "인증에 실패했습니다.")
        }
    }

    // Get current user profile
    suspend fun getCurrentUser(): UserProfile? {
        return try {
            val token = getToken()
            if (token.isNullOrEmpty()) return null
            
            val response = authApi.getCurrentUser("Bearer $token")
            if (response.isSuccessful) {
                response.body()
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    // Logout user
    suspend fun logout() {
        try {
            val token = getToken()
            if (!token.isNullOrEmpty()) {
                authApi.logout("Bearer $token")
            }
        } catch (e: Exception) {
            // Ignore logout errors
        } finally {
            clearAuthData()
        }
    }

    // Check if user is authenticated
    suspend fun isAuthenticated(): Boolean {
        val token = getToken()
        if (token.isNullOrEmpty()) return false
        
        return try {
            getCurrentUser() != null
        } catch (e: Exception) {
            false
        }
    }

    // Get stored auth token
    suspend fun getToken(): String? {
        return context.dataStore.data.map { preferences ->
            preferences[TOKEN_KEY]
        }.first()
    }

    // Get stored user profile
    suspend fun getStoredUser(): UserProfile? {
        return try {
            val userJson = context.dataStore.data.map { preferences ->
                preferences[USER_KEY]
            }.first()
            
            if (userJson != null) {
                gson.fromJson(userJson, UserProfile::class.java)
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    // Get auth state flow
    fun getAuthState(): Flow<Boolean> {
        return context.dataStore.data.map { preferences ->
            preferences[TOKEN_KEY] != null
        }
    }

    // Save authentication data
    private suspend fun saveAuthData(token: String, userProfile: UserProfile) {
        context.dataStore.edit { preferences ->
            preferences[TOKEN_KEY] = token
            preferences[USER_KEY] = gson.toJson(userProfile)
        }
    }

    // Clear authentication data
    private suspend fun clearAuthData() {
        context.dataStore.edit { preferences ->
            preferences.remove(TOKEN_KEY)
            preferences.remove(USER_KEY)
        }
    }
}